precompiles/erc20/approve_test.go

```diff
@@ -0,0 +1,562 @@
+func (s *PrecompileTestSuite) TestDecreaseAllowance() {
+	method := s.precompile.Methods[authorization.DecreaseAllowanceMethod]
+	amount := int64(100)
+	decreaseAmount := int64(50)
+
+	testcases := []struct {
+		name        string
+		malleate    func() []interface{}
+		postCheck   func()
+		expPass     bool
+		errContains string
+	}{
+		{
+			name:        "fail - empty args",
+			malleate:    func() []interface{} { return nil },
+			errContains: "invalid number of arguments",
+		},
+		{
+			name: "fail - invalid number of arguments",
+			malleate: func() []interface{} {
+				return []interface{}{
+					1, 2, 3,
+				}
+			},
+			errContains: "invalid number of arguments",
+		},
+		{
+			name: "fail - invalid address",
+			malleate: func() []interface{} {
+				return []interface{}{
+					"invalid address", big.NewInt(2),
+				}
+			},
+			errContains: "invalid address",
+		},
+		{
+			name: "fail - invalid amount",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), "invalid amount",
+				}
+			},
+			errContains: "invalid amount",
+		},
+		{
+			name: "fail - negative amount",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(-1),
+				}
+			},
+			errContains: "cannot decrease allowance with non-positive values",
+		},
+		{
+			name: "fail - decrease allowance without existing authorization",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(decreaseAmount),
+				}
+			},
+			errContains: "does not exist or is expired",
+		},
+		{
+			name: "pass - decrease allowance with existing authorization",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(decreaseAmount),
+				}
+			},
+			expPass: true,
+			postCheck: func() {
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, decreaseAmount)),
+					[]string{},
+				)
+			},
+		},
+		{
+			// TODO: do we want this to fail or should it delete the authorization? currently fails
+			name: "fail - decrease to zero and delete existing authorization",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(amount),
+				}
+			},
+			errContains: "spend limit must be positive",
+			postCheck: func() {
+				// NOTE: since the authorization is not deleted, we check that it still exists
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+					[]string{},
+				)
+			},
+		},
```

Comment:
I would have expected the authorization to be deleted when reducing the allowance to zero, but it fails currently.

Reply:
that's a bug